`ColorPalette`: refine custom color button's label

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Change the label for the custom color button in the `ColorPalette` component:

- format the color value to always be upper case, and hide the starting `#` character
- display the color name

<details>

<summary>See more details</summary>

For reference, this is how this button looks like in `trunk`

![image (1)](https://user-images.githubusercontent.com/1083581/157872196-f4909cc3-8309-4407-abd2-4a175294b9ed.png)

And this is how it _should_ look like:

![image](https://user-images.githubusercontent.com/1083581/157872416-0a80dc50-d8ac-4c8f-875f-3a2ac7e30595.png)
</details>

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Iterating and refining our UIs

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

1. Added a utility function that is able to find the color name, given its hex value, by matching it against the color palette(s)
2. Used the `Flex` and `Truncate` components (and a couple of custom styles) to implement the layout changes
3. Added a fix to `fieldset` to make sure that it does't grow larger than its container.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

- Check the component in Storybook / Global Styles (in the theme editor) / Post editor (e.g. by setting a background color for a group block)
- Focus your attention on the custom color button, in the `ColorPalette` component:
    - The color value should be uppercase and stripped from the leading `#` character
    - The color name should be displayed. If the current color can't be matched against the current paletted of predefined named colors, the text `Custom` should be displayed instead
    - Make sure that, in case of a color with a very long name, the text is truncated correctly.

## Screenshots or screencast <!-- if applicable -->

**In Storybook:**

https://user-images.githubusercontent.com/1083581/157878761-41d43388-b72b-4ed1-9921-95815e1e3c77.mp4

**In the theme editor:**

https://user-images.githubusercontent.com/1083581/157874005-6f997ffc-43d3-4b0c-894e-12114aa88a6f.mp4

**In the post editor:**

https://user-images.githubusercontent.com/1083581/157875711-94cb9db8-7f80-4f8e-99ce-0b9d2d34cd60.mp4

ciampo committed:
`ColorPalette`: refine custom color button's label

ciampo committed:
Fix `fieldset` growing too wide

ciampo committed:
Transform color value to uppercase via styles

ciampo committed:
Remove unnecessary styles

ciampo committed:
Improve check to match a color to its corresponding one in the palette

ciampo committed:
Memoize button label name computation

ciampo committed:
CHANGELOG

ciampo commented on packages/components/src/color-palette/index.js:
```diff
@@ -125,6 +127,56 @@ export function CustomColorPickerDropdown( { isRenderedInSidebar, ...props } ) {
+					colord( currentValue ).toHex() ===
+					colord( colorValue ).toHex()
+				) {
+					return sprintf(
+						// translators: %1$s: The name of the color e.g: "vivid red"; %2$s: the color palette name (e.g. 'Theme', 'Default', 'Custom'...).
+						__( '%1$s (%2$s)' ),
+						colorName,
+						colorPaletteName
+					);
```
Not sure what to when multiple palettes are passed to the `ColorPalette` component.

For now, i've decided to show the `colorPaletteName` (i.e. `Theme`, `Default`, `Custom`) in between parenthesis after the color name. This was mainly triggered by the fact that `Theme` colors have names like "primary", "secondary", "background"...

Other options could be:

- do not display the `colorPaletteName`
- ignore colors coming from the `Theme` palette when trying to match them with a name (although, are we guaranteed that the theme palette is always going to be called `Theme`? Will it change in other locales?)

ciampo committed:
No need for translations

ciampo commented on packages/components/src/color-palette/index.js:
```diff
@@ -125,6 +127,50 @@ export function CustomColorPickerDropdown( { isRenderedInSidebar, ...props } ) {
+					colord( currentValue ).toHex() ===
+					colord( colorValue ).toHex()
+				) {
+					return `${ colorName }${
+						colorPaletteName ? ` (${ colorPaletteName })` : ''
+					}`;
```
Not sure what to when multiple palettes are passed to the `ColorPalette` component.

For now, i've decided to show the `colorPaletteName` (i.e. `Theme`, `Default`, `Custom`) in between parenthesis after the color name. This was mainly triggered by the fact that `Theme` colors have names like "primary", "secondary", "background"...

Other options could be:

- do not display the `colorPaletteName`
- ignore colors coming from the `Theme` palette when trying to match them with a name (although, are we guaranteed that the theme palette is always going to be called `Theme`? Will it change in other locales?)

mtias replied:
It's good to show the name, it helps convey that themes should not be using names like "Background" which are really Elements and not palette items.

If the color that is set is a custom one without a name we should say "Custom"

ciampo commented on packages/components/src/color-palette/index.js:
```diff
@@ -125,6 +127,50 @@ export function CustomColorPickerDropdown( { isRenderedInSidebar, ...props } ) {
+					colord( currentValue ).toHex() ===
+					colord( colorValue ).toHex()
+				) {
+					return `${ colorName }${
+						colorPaletteName ? ` (${ colorPaletteName })` : ''
+					}`;
```
Is there a need to mark this text for translation?

ciampo committed:
Rename variable

ciampo committed:
Update unit tests

ciampo commented on packages/components/src/color-palette/test/index.js:
```diff
@@ -92,7 +92,7 @@ describe( 'ColorPalette', () => {
 			const onToggle = jest.fn();
 
-			const renderedToggleButton = shallow(
+			const renderedToggleButton = mount(
```
Unit tests were failing, I believe because the switch from a vanilla `button` to a `Flex as={button}` component meant that Enzyme's shallow render couldn't find that `button` anymore. Changing to `mount` solves the problems (with the consequence of generating a larger snapshot).

Not sure if this is the cleanest solution, but I figured it'd be good enough — especially considering that we should eventually rewrite these tests to use `@testing-library` instead.

jasmussen commented:
Nice one. This is what I see:
<img width="607" alt="Screenshot 2022-03-11 at 15 30 28" src="https://user-images.githubusercontent.com/1204802/157887145-d0a92989-96e8-4cad-82d9-5cf75d021b14.png">

I'll take a look and see if I can't get the padding to match this:

<img width="277" alt="Screenshot 2022-03-11 at 15 31 23" src="https://user-images.githubusercontent.com/1204802/157887249-03a4e70f-8dab-42d8-b443-15209e9814d3.png">

Do we need to include the (Theme)? I feel like there's plenty of context in the sectioning off of the swatches.

mtias commented:
Yes, let's try without the (Theme)

jasmussen commented:
For the spacings, here's the source design:
<img width="692" alt="Screenshot 2022-03-11 at 15 43 33" src="https://user-images.githubusercontent.com/1204802/157889753-6ec0ba71-3cc0-4964-9f9b-a24d23e7d71c.png">

<img width="741" alt="Screenshot 2022-03-11 at 15 44 24" src="https://user-images.githubusercontent.com/1204802/157889794-eabad730-f5cd-4a28-a18f-a23b60030423.png">

The card is 64px tall, and the red padding as denoted is 12px.

The card inside the flyout is a bit wider, that's okay, the height and padding should still be the same:
<img width="424" alt="Screenshot 2022-03-11 at 15 45 53" src="https://user-images.githubusercontent.com/1204802/157890016-08ee1fa5-72eb-45a6-9769-4aa195c53d89.png">

ciampo committed:
Style tweaks (height, padding, vertical alignment)

ciampo committed:
Do not show color palette's name

ciampo committed:
Update snapshots

ciampo commented:
Thank you for the quick round of feedback!

I went ahead and implemented it:

1. Tweaked styles (height, padding, vertical alignment) https://github.com/WordPress/gutenberg/pull/39386/commits/48b3822c26d6089df7447ff844df862a0896c8b2
2. Removed color palette name https://github.com/WordPress/gutenberg/pull/39386/commits/4f98c7cdc069707a10329e105f8f86ded96f09a7


<details>

<summary>A few screenshots:</summary>


![Screenshot 2022-03-11 at 16 08 52](https://user-images.githubusercontent.com/1083581/157893916-7c8cff26-5869-4ae5-a56f-c445bd8784cf.png)

![Screenshot 2022-03-11 at 16 09 40](https://user-images.githubusercontent.com/1083581/157894028-1ea37ac9-52d2-44c9-bdfa-364fbe088bab.png)

![Screenshot 2022-03-11 at 16 09 58](https://user-images.githubusercontent.com/1083581/157894101-b6fb1c0d-36d1-4321-8a8a-2380a1faf017.png)

</details>

ciampo commented:
Also flagging some failing e2e tests that seem related to the changes in this PR.

I'll still have to wait until next week to merge, since currently e2e tests are broken across all PRs.

jasmussen commented:
Nice, the metrics are right now!

<img width="642" alt="Screenshot 2022-03-11 at 16 09 04" src="https://user-images.githubusercontent.com/1204802/157893983-240233f3-abc1-4ce2-90d1-d9f0d54d0e34.png">

For that reason, I'm happy to give this a thumbs up. 

The unset state looks odd now, though — it seems like the previous height was informed by the transparency grid size:
<img width="692" alt="Screenshot 2022-03-11 at 16 09 00" src="https://user-images.githubusercontent.com/1204802/157894093-8bfbb9e9-f0ba-4c64-a0a8-e0c16e5a6fbd.png">

Fundamentally I consider that a separate problem, one that is tracked in #36998, and so it probably doesn't need to be fixed by this PR. But it does feel like something to address sooner rather than later.

ciampo committed:
Attempt to fix e2e test

mirka commented on packages/components/src/color-palette/index.js:
```diff
@@ -125,6 +127,45 @@ export function CustomColorPickerDropdown( { isRenderedInSidebar, ...props } ) {
+		return '';
+	}
+
+	let colorNameToReturn;
```
Did you originally intend to assign a value to this somewhere? `colorNameToReturn` is never set in this code block.

ciampo replied:
Good point — That was a left-over of a refactor 😅 

Clean up in https://github.com/WordPress/gutenberg/pull/39386/commits/0e585aded465aea7a0bd48ed65583c30d0f1dedf

mirka commented on packages/components/src/color-palette/index.js:
```diff
@@ -125,6 +127,45 @@ export function CustomColorPickerDropdown( { isRenderedInSidebar, ...props } ) {
+
+	let colorNameToReturn;
+
+	if ( showMultiplePalettes ) {
+		for ( const { colors: colorPaletteColors } of colors ) {
```
Non-blocking nit: Because the body of these loops are the same in both conditions, I wonder if we should normalize the arrays earlier to simplify the code? For example:

```js
const colorPalettes = showMultiplePalettes ? colors : [ { colors } ];

for ( const { colors: colorPaletteColors } of colorPalettes ) {
	for ( const {
		name: colorName,
		color: colorValue,
	} of colorPaletteColors ) {
		if (
			colord( currentValue ).toHex() ===
			colord( colorValue ).toHex()
		) {
			return colorName;
		}
	}
}
```

ciampo replied:
Such a neat idea! Applied in https://github.com/WordPress/gutenberg/pull/39386/commits/73defe321e1dd5c0e54af422a2cf59a7cbb4a9c1

mirka commented on packages/components/CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 
+### Enhancements
+
+-   `ColorPalette`: : refine custom color button's label. ([#39386](https://github.com/WordPress/gutenberg/pull/39386))
```
```suggestion
-   `ColorPalette`: refine custom color button's label. ([#39386](https://github.com/WordPress/gutenberg/pull/39386))
```

mirka commented on packages/components/src/color-palette/index.js:
```diff
@@ -165,7 +218,10 @@ export default function ColorPalette( {
 					renderToggle={ ( { isOpen, onToggle } ) => (
-						<button
+						<Flex
+							as={ 'button' }
```
Kind of similar to [this comment](https://github.com/WordPress/gutenberg/pull/37769#discussion_r815799705) — should the button's aria-label include the current color name and value?

ciampo replied:
I gave your idea a try in https://github.com/WordPress/gutenberg/pull/39386/commits/ba6bdd436e0d117bf53a23fd214a7e574bb00e69 and https://github.com/WordPress/gutenberg/pull/39386/commits/bbe98743ef11d53a0dcbb1e52820a2c760f1d53d — although I'm not 10)% sure about the wording that I chose.

Maybe @alexstine can give some feedback on whether the new `aria-label` needs any further improvements?

alexstine replied:
@ciampo Are you able to make the button contain what the default color is without any colors selected?

I am not sure why the color picker opens much lower in the DOM. It used to not do this if memory serves me right but this PR didn't change that. Chocking that up to a regression that happened at some point. This is going to need a considerable amount of work after this gets merged. Something else to focus on.

ciampo replied:
> Are you able to make the button contain what the default color is without any colors selected?

I'm not sure I understand what you're exactly referring to. When no color is "picked", I believe that the `value` prop is either `undefined` (or an empty string) — in other words, this component doesn't have a "default" color. At the moment, when the `value` prop is not defined, the aria-label for this `button` should be "Custom color picker."

> I am not sure why the color picker opens much lower in the DOM. It used to not do this if memory serves me right but this PR didn't change that. Chocking that up to a regression that happened at some point. This is going to need a considerable amount of work after this gets merged. Something else to focus on.

I haven't really worked on this component before, but I'm going to ping @jorgefilipecosta and @ntsekouras in case they have a better understanding of the matter that you're highlighting here.

alexstine replied:
@ciampo I figured the component had a default value when no color was selected. If not, this PR is good to go on accessibility end.

mirka reviewed (commented):
Looks very nice ✨ 

One tiny thing: In the Storybook, the mock color data has lowercase `name`s. Just as a stylistic thing, could we capitalize those mock color names? (I hope most of the color data in actual usage already have capitalized names. Otherwise we need to consider a text-transform!)

ntsekouras commented on packages/base-styles/_variables.scss:
```diff
@@ -37,7 +37,8 @@ $grid-unit-30: 3 * $grid-unit;		// 24px
 $grid-unit-50: 5 * $grid-unit;		// 40px
 $grid-unit-60: 6 * $grid-unit;		// 48px
-
+$grid-unit-70: 7 * $grid-unit;    // 56px
```
Nit: spaces are not consistent.

ciampo replied:
Well spotted! I swapped spaces with tabs in https://github.com/WordPress/gutenberg/pull/39386/commits/1a89543ba69bd6d7461c3831e980efdee0aefc86

ntsekouras reviewed (approved):
This looks so good Marco! Thanks! 💯 

I'm approving this but I think @mirka's comments about the unused `colorNameToReturn` and the normalization of input would help to simplify the code.

ciampo committed:
Avoid using `colorNameToReturn` variable

ciampo committed:
Simplify colorname extraction loop by normalising input array's format

ciampo committed:
Add color name and value to button's aria label

ciampo committed:
Use tabs instead of spaces in `variables.scss`

ciampo committed:
Update packages/components/CHANGELOG.md

Co-authored-by: Lena Morita <lena@jaguchi.com>

ciampo committed:
Fallback aria label if value is not defined

ciampo committed:
Update snapshots

ciampo commented:
The [`Heading › should correctly apply custom colors` test](https://github.com/WordPress/gutenberg/blob/08759a972d74b214853ebc23267cbc9f6db41c6b/packages/e2e-tests/specs/editor/blocks/heading.test.js#L91) has been failing consistently on this PR, although when I test it locally it keeps passing.

Any ideas about why this could happen?

alexstine commented:
@ciampo Try to make your environment as close to the real thing as possible.

1. npm ci
2. npm run build
3. npx wp-env start
4. npm run test-e2e -- /path/to/test/

Sometimes I think E2E tests have a mind of their own. Also, you may want to try to run it a couple times. Possible that GitHub Actions server is just running slow. Every now and then, I get green checks from doing nothing. 👍 

I'll accessibility review this later today.

ntsekouras committed:
fix e2e test

ntsekouras commented:
>The [Heading › should correctly apply custom colors test](https://github.com/WordPress/gutenberg/blob/08759a972d74b214853ebc23267cbc9f6db41c6b/packages/e2e-tests/specs/editor/blocks/heading.test.js#L91) has been failing consistently on this PR, although when I test it locally it keeps passing.

> Any ideas about why this could happen?

For me it was failing locally as well. Now the `button` has no text nodes, so it seems it has to do with internals of xpath and how handles nested nodes. Here is [something I read](https://stackoverflow.com/questions/3655549/xpath-containstext-some-string-doesnt-work-when-used-with-node-with-more) that helped a bit.

ciampo commented:
Thank you both! Thanks to @alexstine 's notes, with `npm ci && npm run build` I managed to replicate the failure, which was then addressed by @ntsekouras 's fix (I wrongly assumed that `text()` and `contains()` would be able to read text content from nested nodes 😅 )

alexstine reviewed (approved):
Functionality and accessibility seem sane to me. 👍 Thanks @ciampo !

ciampo committed:
Refactor story to controls

ciampo commented:
Update:

- Rebased to solve CHANGELOG conflicts
- Also [updated Storybook to controls](https://github.com/WordPress/gutenberg/pull/39386/commits/0c53767ba2fe9179b085c13bf81bd589ab009f03), while including latest piece of feedback from @mirka (about capitalising color names)

E2E tests flakiness seems to be solved, so hopefully I'll be able to merge once all CI checks complete

mtias commented:
Thank you very much for the refinement and attention to detail!
